sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,373 @@
+        expiry = time.time() + 3600  # Default to 1 hour.
+    encoded_uri = quote_plus(uri)
+    ttl = int(expiry)
+    sign_key = '{}\n{}'.format(encoded_uri, ttl)
+    signature = b64encode(HMAC(b64decode(key), sign_key.encode('utf-8'), sha256).digest())
+    result = {
+        'sr': uri,
+        'sig': signature,
+        'se': str(ttl)}
+    if policy:
+        result['skn'] = policy
+    return 'SharedAccessSignature ' + urlencode(result)
+
+
+def _build_uri(address, entity):
+    parsed = urlparse(address)
+    if parsed.path:
+        return address
+    if not entity:
+        raise ValueError("No EventHub specified")
+    address += "/" + str(entity)
+    return address
+
+
+_Address = collections.namedtuple('Address', 'hostname path')
+
+
+class ClientBase(object):  # pylint:disable=too-many-instance-attributes
+    def __init__(self, host, event_hub_path, credential, **kwargs):
+        self.eh_name = event_hub_path
+        path = "/" + event_hub_path if event_hub_path else ""
+        self._address = _Address(hostname=host, path=path)
+        self._container_id = CONTAINER_PREFIX + str(uuid.uuid4())[:8]
+        self._credential = credential
+        self._keep_alive = kwargs.get("keep_alive", 30)
+        self._auto_reconnect = kwargs.get("auto_reconnect", True)
+        self._mgmt_target = "amqps://{}/{}".format(self._address.hostname, self.eh_name)
+        self._auth_uri = "sb://{}{}".format(self._address.hostname, self._address.path)
+        self._config = Configuration(**kwargs)
+        self._debug = self._config.network_tracing
+        self._conn_manager = get_connection_manager(**kwargs)
+        _LOGGER.info("%r: Created the Event Hub client", self._container_id)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    @classmethod
+    def from_connection_string(cls, conn_str, **kwargs):
+        event_hub_path = kwargs.pop("event_hub_path", None)
+        address, policy, key, entity = _parse_conn_str(conn_str)
+        entity = event_hub_path or entity
+        left_slash_pos = address.find("//")
+        if left_slash_pos != -1:
+            host = address[left_slash_pos + 2:]
+        else:
+            host = address
+        return cls(host, entity, EventHubSharedKeyCredential(policy, key), **kwargs)
+
+    def _create_auth(self):
+        """
+        Create an ~uamqp.authentication.SASTokenAuth instance to authenticate
+        the session.
+        """
+        http_proxy = self._config.http_proxy
+        transport_type = self._config.transport_type
+        auth_timeout = self._config.auth_timeout
+
+        # TODO: the following code can be refactored to create auth from classes directly instead of using if-else
+        if isinstance(self._credential, EventHubSharedKeyCredential):  # pylint:disable=no-else-return
+            username = self._credential.policy
+            password = self._credential.key
+            if "@sas.root" in username:
```

Comment:
Does this need to be case insensitive?

Reply:
You can ignore the `_create_auth` code - this will be refactored separately in a subsequent PR (that will remove all the str crap ;) )